usr/sbin/pkcstok_migrate/pkcstok_migrate.c

```diff
@@ -0,0 +1,2367 @@
+    if (userpin)
+        printf("  user PIN specified\n");
+    if (sopin)
+        printf("  SO PIN specified\n");
+    if (vlevel >= 0) {
+        trace_level = vlevel;
+        printf("  verbose level = %s\n", verbose);
+    }
+    printf("\n");
+
+    /* Slot ID must be given */
+    if (slot_id == 0) {
+        printf("Slot ID must be specified.\n");
+        goto done;
+    }
+
+    /* Datastore must be given */
+    if (data_store == NULL) {
+        printf("Data store path must be specified.\n");
+        goto done;
+    }
+
+    /* Location of opencryptoki.conf must be specified. */
+    if (conf_dir == NULL) {
+        printf("Location of opencryptoki.conf must be specified.\n");
+        goto done;
+    }
+
+    /* Check if given data_store exists */
+    if (!datastore_exists(data_store)) {
+        ret = CKR_FUNCTION_FAILED;
+        printf("Datastore %s does not exist.\n", data_store);
+        goto done;
+    }
+
+    /* Check if given conf_dir exists and contains opencryptoki.conf */
+    if (!conffile_exists(conf_dir)) {
+        ret = CKR_FUNCTION_FAILED;
+        printf("%s does not exist or does not contain opencryptoki.conf.\n", conf_dir);
+        goto done;
+    }
+
+    /* Check if there are token objects to migrate */
+    if (datastore_empty(data_store)) {
+        ret = CKR_OK;
+        printf("Datastore %s is empty, no objects to migrate.\n", data_store);
+        goto done;
+    }
+
+    /* Check if pkcsslotd is running */
+    if (pkcsslotd_running()) {
+        printf("Please stop pkcsslotd before running this utility.\n");
+        ret = CKR_OK;
+        goto done;
+    }
+
+    /* Identify token related to given slot ID */
+    ret = identify_token(slot_id, conf_dir, &dll_name);
+    if (ret != CKR_OK) {
+        printf("Cannot identify a token related to given slot ID %ld\n", slot_id);
+        goto done;
+    }
+
+    /* Get token info from NVTOK.DAT */
+    ret = get_token_info(data_store, &tokinfo);
+    if (ret != CKR_OK) {
+        printf("Cannot get the token label from NVTOK.DAT\n");
+        goto done;
+    }
+
+    /* Check with user if ok to migrate this token */
+    printf("Slot ID %ld points to DLL name %s, which is a %s token.\n",
+           slot_id, dll_name, dll2name(dll_name));
+    printf("Data store %s points to this token info:\n", data_store);
+    printf("  label           : [%s]\n", strip_blanks(tokinfo.label, 32));
+    printf("  manufacturerID  : [%s]\n", strip_blanks(tokinfo.manufacturerID, 32));
+    printf("  model           : [%s]\n", strip_blanks(tokinfo.model, 16));
+    printf("  serialNumber    : [%s]\n", strip_blanks(tokinfo.serialNumber, 16));
+    printf("  hardwareVersion : %i.%i\n", tokinfo.hardwareVersion.major, tokinfo.hardwareVersion.minor);
+    printf("  firmwareVersion : %i.%i\n", tokinfo.firmwareVersion.major, tokinfo.firmwareVersion.minor);
+    printf("Migrate this token with given slot ID? y/n\n");
+    num_chars = getline(&buff, &buflen, stdin);
```

Comment:
where is buff freed ?

Reply:
Good point! The free at the end of main went away due to intermediate code changes. Added it again.